### PR TITLE
feat: [CI-5075]: LE and Addon changes for background step

### PR DESCRIPTION
**Pull request by raghavharness**
## Describe your changes
LE and Addon changes for background step
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/36365)

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Passed_** 
 


            
 No issues Everything looks good
 [Details](https://app.embold.io/organization/gh/harness/repositories/7752dbb3099d2dba3b5d37f5931a5e4c/pull-request/36365)

File: product/ci/addon/tasks/run.go [hunk 1]
```diff
@@ -138,15 +146,24 @@ func (r *runTask) execute(ctx context.Context, retryCount int32) (map[string]str
...
 	}
-	cmdArgs := []string{entrypointArg, cmdToExecute}
 
-	cmd := r.cmdContextFactory.CmdContextWithSleep(ctx, cmdExitWaitTime, shell, cmdArgs...).
-		WithStdout(r.procWriter).WithStderr(r.procWriter).WithEnvVarsMap(envVars)
-	err = runCmd(ctx, cmd, r.id, cmdArgs, retryCount, start, r.logMetrics, r.addonLogger)
+	if r.detach {
+		go func() {
+			ctx := context.Background()
+			cmd := r.cmdContextFactory.CmdContextWithSleep(ctx, cmdExitWaitTime, cmdArgs[0], cmdArgs[1:]...).
```
Comment (shubham149) [inline product/ci/addon/tasks/run.go:157]: Is cmd same for detach or non-detach op? We can move it out if that is true

File: product/ci/addon/tasks/run.go [hunk 2]
```diff
@@ -138,15 +146,24 @@ func (r *runTask) execute(ctx context.Context, retryCount int32) (map[string]str
...
+				WithStdout(r.procWriter).WithStderr(r.procWriter).WithEnvVarsMap(envVars)
+			_ = runCmd(ctx, cmd, r.id, cmdArgs, retryCount, start, r.logMetrics, r.addonLogger)
+		}()
+	} else {
+		cmd := r.cmdContextFactory.CmdContextWithSleep(ctx, cmdExitWaitTime, cmdArgs[0], cmdArgs[1:]...).
+			WithStdout(r.procWriter).WithStderr(r.procWriter).WithEnvVarsMap(envVars)
+		err = runCmd(ctx, cmd, r.id, cmdArgs, retryCount, start, r.logMetrics, r.addonLogger)
+	}
+
```
Comment (shubham149) [inline product/ci/addon/tasks/run.go:166]: How are we ensuring that r.envVarOutputs is of empty size if detach is true?

File: product/ci/addon/grpc/handler.go [hunk 3]
```diff
@@ -67,13 +67,21 @@ func (h *handler) ExecuteStep(ctx context.Context, in *pb.ExecuteStepRequest) (*
 
 	switch x := in.GetStep().GetStep().(type) {
 	case *enginepb.UnitStep_Run:
-		stepOutput, numRetries, err := newRunTask(in.GetStep(), in.GetPrevStepOutputs(), in.GetTmpFilePath(), rl.BaseLogger,
+		var (
```
[the remark this answers is not in the document]
Comment (raghavharness) [reply product/ci/addon/grpc/handler.go:70]: I did this in an earlier implementation. Will change it back.

File: product/ci/addon/tasks/run.go (hunk 1 above)
Comment (raghavharness) [reply product/ci/addon/tasks/run.go:157]: only difference is the ctx value. We need to set a new ctx in this.

File: product/ci/addon/tasks/run.go (hunk 2 above)
Comment (raghavharness) [reply product/ci/addon/tasks/run.go:166]: We set it as null in manager

**Review by shubham149 — COMMENTED** [review]
> Reviewed 2 of 6 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36365)* status: 3 file(s) to review. 0 of 1 approvals obtained.

**Review by shubham149 — DISMISSED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36365#-:-N9erHPU7EEsJDZQ4Fgj:bnfp4nl)* status: 3 file(s) to review.

**Review by shubham149 — APPROVED** [review]
> Reviewed 3 of 6 files at r1, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/36365)* status: Some of the checks are too old.

**Pull request merged**